src/core/queriesObserver.ts

```diff
@@ -64,78 +64,121 @@ export class QueriesObserver extends Subscribable<QueriesObserverListener> {
   }
 
   getOptimisticResult(queries: QueryObserverOptions[]): QueryObserverResult[] {
-    return queries.map((options, index) => {
-      const defaultedOptions = this.client.defaultQueryObserverOptions(options)
-      return this.getObserver(defaultedOptions, index).getOptimisticResult(
-        defaultedOptions
-      )
+    return this.getUpdatedObservers(queries).newResult
+  }
+
+  private getUpdatedObservers(
+    queries: QueryObserverOptions[],
+    notifyOptions?: NotifyOptions
+  ): QueryObserverUpdate {
+    const prevObservers = this.observers
+    const prevObserversMap = this.observersMap
+    const newResult: QueryObserverResult[] = []
+    const newObservers: QueryObserver[] = []
+    const newObserversMap: Record<string, QueryObserver> = {}
+
+    const defaultedQueryOptions = queries.map(options =>
+      this.client.defaultQueryObserverOptions(options)
+    )
+    const matchingObservers = defaultedQueryOptions
+      .map(options => {
+        if (options.queryHash == null) {
+          return null
```

Comment:
codecov is reporting here that his line isn't covered, and I think maybe we can even remove the check and use `options.queryHash!` like before?

judging from the code, we should _always_ have a query hash when working with defaulted options, seems just quite hard to convince the compiler about it.

the hash is added to the options here if it does not yet exist: https://github.com/tannerlinsley/react-query/blob/f21d932142e8618995936c12c5f5696aa8993473/src/core/queryClient.ts#L619-L624

Reply:
Thanks for the comment, I hadn't thought about those filters too much, just assumed empty hashes could be filtered out. 

However, when I tested this out a bit more, I noticed that you actually _can_ pass an undefined `queryKey` via options. This will make the `queryHash` undefined as well. And it does seem to work - somewhat. All the falsy `queryKeys` will just have an undefined `queryHash`, but because Map does allow undefined keys, the same single observer will be returned for all falsy `queryKey` values. 

So assuming someone might be using this, I changed the code to not filter out undefined `queryHash`values, but just use them as keys. I added a couple of tests for this case, so the coverage should be up to 100% again.

I think I'd prefer to use the nullish coalescing operator (`queryHash ?? ''`) to make it more explicit we are allowing for null or undefined hashes here. However, the previous non-null assertion operator (`queryHash!`) would probably also work in this case, allowing undefined to be used in the Map keys. So I can change to that if that is more commonly used here?

Reply:
> I noticed that you actually can pass an undefined queryKey via options

TypeScript types don't allow for it. Sure, if you use JavaScript, you can always pass in anything :)

> However, the previous non-null assertion operator (queryHash!) would probably also work in this case, allowing undefined to be used in the Map keys. So I can change to that if that is more commonly used here?

yes please. I'm usually not a fan of the bang operator, but I also don't like to add code in a library for cases that will not happen if it increases the codebase

Reply:
:+1: Changed the Map indexing to use the bang operator. 

> TypeScript types don't allow for it. Sure, if you use JavaScript, you can always pass in anything :)

I think you can pass an undefined `queryKey` also in TypeScript, if you only use the QueryOptions object for the parameters, apparently because there [queryKey is marked as optional](https://github.com/tannerlinsley/react-query/blob/master/src/core/types.ts#L69). E.g. `useQueries([{ queryKey: undefined, queryFn: () => 3 }])` seems to compile and run.

But true, also JS allows you to pass in anything, so I guess we need to support it anyways :)

Reply:
It will likely become required in v4 :)